### ISSUE TITLE
High resident set size (RSS)

### ISSUE THREAD

**Issue by wks** [mmtk/mmtk-core#1220]
We observed on several VM bindings that the RSS memory consumption is higher when using MMTk compared to those VMs' default GCs.  We need to inspect where those memory is used for.  Possibilities include (but are not limited to)

-   The SFT
-   The use of `BlockQueue` which caches memory blocks instead of returning the memory to the OS by unmapping.
-   Auxillary Rust data structures, such as `Vec`, `Box`, etc.

**Comment by qinsoon** [comment]
I tried to log RSS usage during a `fop` run with our OpenJDK binding. The large RSS issue is mostly caused by the initialization of SFT and VMMap.

Before MMTk initializes, the RSS is 14M. After MMTk initializes `SFT_MAP`, the RSS is 526MB. After MMTk initializes `VM_MAP`, the RSS is 783MB. After the plan is created, the RSS is 785MB. After MMTk finishes the initialization, the RSS is 801MB. At the end of the `fop` run, the RSS is 900MB+.

Though it is worth investigating how the RSS grows over the run, we should focus on SFT and VM map first. They in total contribute to 80% of the RSS usage.

**Comment by wks** [comment]
@qinsoon Can you also evaluated the effect of

-   heap size (and with dynamic heap size, the max heap size)
-   Map32 vs Map64

and see whether they affect the RSS impact of `SFT_MAP` and `VM_MAP`, too?  I think they affect the heap layout and may also have an effect on the memory `SFT_MAP` and `VM_MAP` try to mmap.

**Comment by qinsoon** [comment]
I was using a compressed pointer build. So it used `SFTSparseChunkMap` and `Map32`.

**Comment by k-sareen** [comment]
Yes. I ran tests ages ago and I mentioned back then that we've regressed because the restricted address space uses the [sparse chunk map](https://github.com/mmtk/mmtk-core/pull/632). That, VMMap, and not returning pages back to the OS were the largest sources of RSS overhead from memory

**Comment by qinsoon** [comment]
Right. Without compressed pointer (using `SFTSpaceMap` and `Map64`), there is no substantial RSS increase during initialization. At the end of initilaization, the RSS was 17MB. During the `fop` run, the RSS increased from 43MB (first GC) to 221MB (last GC).

On the contrary, with compressed pointer (using `SFTSparseMap` and `Map32`), the RSS was 801MB and during the run, the RSS increased from 833MB (first GC) to 1023MB (last GC).
